<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
<meta name="viewport" content="width=device-width, initial-scale=1, maximum-scale=2">
<meta name="theme-color" content="#222">
<meta name="generator" content="Hexo 4.2.0">
  <link rel="apple-touch-icon" sizes="180x180" href="/codeHub/uploads/themeImg/logo.png">
  <link rel="icon" type="image/png" sizes="32x32" href="/codeHub/uploads/themeImg/logo.png">
  <link rel="icon" type="image/png" sizes="16x16" href="/codeHub/uploads/themeImg/logo.png">
  <link rel="mask-icon" href="/codeHub/uploads/themeImg/logo.png" color="#222">

<link rel="stylesheet" href="/codeHub/css/main.css">

<link rel="stylesheet" href="//fonts.googleapis.com/css?family=Lato:300,300italic,400,400italic,700,700italic&display=swap&subset=latin,latin-ext">
<link rel="stylesheet" href="/codeHub/lib/font-awesome/css/font-awesome.min.css">
  <link rel="stylesheet" href="//cdn.jsdelivr.net/gh/fancyapps/fancybox@3/dist/jquery.fancybox.min.css">

<script id="hexo-configurations">
    var NexT = window.NexT || {};
    var CONFIG = {"hostname":"zpliu1126.github.io","root":"/codeHub/","scheme":"Gemini","version":"7.7.2","exturl":false,"sidebar":{"position":"left","display":"post","padding":18,"offset":12,"onmobile":false},"copycode":{"enable":true,"show_result":true,"style":null},"back2top":{"enable":true,"sidebar":true,"scrollpercent":true},"bookmark":{"enable":true,"color":"#222","save":"auto"},"fancybox":true,"mediumzoom":false,"lazyload":true,"pangu":false,"comments":{"style":"tabs","active":null,"storage":true,"lazyload":true,"nav":null},"algolia":{"hits":{"per_page":10},"labels":{"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}","hits_stats":"${hits} results found in ${time} ms"}},"localsearch":{"enable":true,"trigger":"auto","top_n_per_article":1,"unescape":false,"preload":false},"motion":{"enable":true,"async":false,"transition":{"post_block":"fadeIn","post_header":"slideDownIn","post_body":"slideDownIn","coll_header":"slideLeftIn","sidebar":"slideUpIn"}},"path":"search.xml"};
  </script>

  <meta name="description" content="差异表达基因​    生物学上不同样本之间的表达差异时服从负二项分布的，RNA-seq中得到的基因表达水平是抽样过程中的一种离散形式。在测得的reads总量一定的情况下，表达水平越高的基因在抽样过程中所占的比例就越高，有些低表达的基因也有可能无法被检测出来。在得到基因的表达量之后，根据实验设计对不同样本之间基因进行差异表达分析  同物种、不同组织间的比较 同一物种、同一组织、在不同处理下的比较 同">
<meta property="og:type" content="article">
<meta property="og:title" content="基因差异表达分析">
<meta property="og:url" content="https://zpliu1126.github.io/codeHub/2020/03/22/02%E5%9F%BA%E5%9B%A0%E5%B7%AE%E5%BC%82%E8%A1%A8%E8%BE%BE%E5%88%86%E6%9E%90/index.html">
<meta property="og:site_name" content="Code Hub">
<meta property="og:description" content="差异表达基因​    生物学上不同样本之间的表达差异时服从负二项分布的，RNA-seq中得到的基因表达水平是抽样过程中的一种离散形式。在测得的reads总量一定的情况下，表达水平越高的基因在抽样过程中所占的比例就越高，有些低表达的基因也有可能无法被检测出来。在得到基因的表达量之后，根据实验设计对不同样本之间基因进行差异表达分析  同物种、不同组织间的比较 同一物种、同一组织、在不同处理下的比较 同">
<meta property="og:locale" content="en_US">
<meta property="og:image" content="https://upload-images.jianshu.io/upload_images/6049898-721e9cbe006abbe3.png">
<meta property="og:image" content="https://s1.ax1x.com/2020/03/17/8t6jXV.png">
<meta property="article:published_time" content="2020-03-22T05:59:13.000Z">
<meta property="article:modified_time" content="2020-03-22T06:03:41.476Z">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="https://upload-images.jianshu.io/upload_images/6049898-721e9cbe006abbe3.png">

<link rel="canonical" href="https://zpliu1126.github.io/codeHub/2020/03/22/02%E5%9F%BA%E5%9B%A0%E5%B7%AE%E5%BC%82%E8%A1%A8%E8%BE%BE%E5%88%86%E6%9E%90/">


<script id="page-configurations">
  // https://hexo.io/docs/variables.html
  CONFIG.page = {
    sidebar: "",
    isHome : false,
    isPost : true,
    lang   : 'en'
  };
</script>

  <title>基因差异表达分析 | Code Hub</title>
  






  <noscript>
  <style>
  .use-motion .brand,
  .use-motion .menu-item,
  .sidebar-inner,
  .use-motion .post-block,
  .use-motion .pagination,
  .use-motion .comments,
  .use-motion .post-header,
  .use-motion .post-body,
  .use-motion .collection-header { opacity: initial; }

  .use-motion .site-title,
  .use-motion .site-subtitle {
    opacity: initial;
    top: initial;
  }

  .use-motion .logo-line-before i { left: initial; }
  .use-motion .logo-line-after i { right: initial; }
  </style>
</noscript>

</head>

<body itemscope itemtype="http://schema.org/WebPage">
  <div class="container use-motion">
    <div class="headband"></div>

    <header class="header" itemscope itemtype="http://schema.org/WPHeader">
      <div class="header-inner"><div class="site-brand-container">
  <div class="site-nav-toggle">
    <div class="toggle" aria-label="Toggle navigation bar">
      <span class="toggle-line toggle-line-first"></span>
      <span class="toggle-line toggle-line-middle"></span>
      <span class="toggle-line toggle-line-last"></span>
    </div>
  </div>

  <div class="site-meta">

    <a href="/codeHub/" class="brand" rel="start">
      <span class="logo-line-before"><i></i></span>
      <h1 class="site-title">Code Hub</h1>
      <span class="logo-line-after"><i></i></span>
    </a>
      <p class="site-subtitle" itemprop="description">keeping learn</p>
  </div>

  <div class="site-nav-right">
    <div class="toggle popup-trigger">
        <i class="fa fa-search fa-fw fa-lg"></i>
    </div>
  </div>
</div>




<nav class="site-nav">
  <ul id="menu" class="menu">
        <li class="menu-item menu-item-首页">

    <a href="/codeHub/" rel="section"><i class="fa fa-fw fa-home"></i>首页</a>

  </li>
        <li class="menu-item menu-item-文章">

    <a href="/codeHub/archives/" rel="section"><i class="fa fa-fw fa-archive"></i>文章<span class="badge">3</span></a>

  </li>
        <li class="menu-item menu-item-分类">

    <a href="/codeHub/categories" rel="section"><i class="fa fa-fw fa-th"></i>分类<span class="badge">2</span></a>

  </li>
        <li class="menu-item menu-item-标签">

    <a href="/codeHub/tags/" rel="section"><i class="fa fa-fw fa-tags"></i>标签<span class="badge">2</span></a>

  </li>
      <li class="menu-item menu-item-search">
        <a role="button" class="popup-trigger"><i class="fa fa-search fa-fw"></i>Search
        </a>
      </li>
  </ul>
</nav>



  <div class="search-pop-overlay">
    <div class="popup search-popup">
        <div class="search-header">
  <span class="search-icon">
    <i class="fa fa-search"></i>
  </span>
  <div class="search-input-container">
    <input autocomplete="off" autocorrect="off" autocapitalize="off"
           placeholder="Searching..." spellcheck="false"
           type="search" class="search-input">
  </div>
  <span class="popup-btn-close">
    <i class="fa fa-times-circle"></i>
  </span>
</div>
<div id="search-result">
  <div id="no-result">
    <i class="fa fa-spinner fa-pulse fa-5x fa-fw"></i>
  </div>
</div>

    </div>
  </div>

</div>
    </header>

    
  <div class="reading-progress-bar"></div>
  <a role="button" class="book-mark-link book-mark-link-fixed"></a>

  <a href="https://github.com/zpliu1126" class="github-corner" title="Follow me on GitHub" aria-label="Follow me on GitHub" rel="noopener" target="_blank"><svg width="200" height="200" viewBox="0 0 250 250" aria-hidden="true"><path d="M0,0 L115,115 L130,115 L142,142 L250,250 L250,0 Z"></path><path d="M128.3,109.0 C113.8,99.7 119.0,89.6 119.0,89.6 C122.0,82.7 120.5,78.6 120.5,78.6 C119.2,72.0 123.4,76.3 123.4,76.3 C127.3,80.9 125.5,87.3 125.5,87.3 C122.9,97.6 130.6,101.9 134.4,103.2" fill="currentColor" style="transform-origin: 130px 106px;" class="octo-arm"></path><path d="M115.0,115.0 C114.9,115.1 118.7,116.5 119.8,115.4 L133.7,101.6 C136.9,99.2 139.9,98.4 142.2,98.6 C133.8,88.0 127.5,74.4 143.8,58.0 C148.5,53.4 154.0,51.2 159.7,51.0 C160.3,49.4 163.2,43.6 171.4,40.1 C171.4,40.1 176.1,42.5 178.8,56.2 C183.1,58.6 187.2,61.8 190.9,65.4 C194.5,69.0 197.7,73.2 200.1,77.6 C213.8,80.2 216.3,84.9 216.3,84.9 C212.7,93.1 206.9,96.0 205.4,96.6 C205.1,102.4 203.0,107.8 198.3,112.5 C181.9,128.9 168.3,122.5 157.7,114.1 C157.9,116.9 156.7,120.9 152.7,124.9 L141.0,136.5 C139.8,137.7 141.6,141.9 141.8,141.8 Z" fill="currentColor" class="octo-body"></path></svg></a>


    <main class="main">
      <div class="main-inner">
        <div class="content-wrap">
          

          <div class="content">
            

  <div class="posts-expand">
      
  
  
  <article itemscope itemtype="http://schema.org/Article" class="post-block " lang="en">
    <link itemprop="mainEntityOfPage" href="https://zpliu1126.github.io/codeHub/2020/03/22/02%E5%9F%BA%E5%9B%A0%E5%B7%AE%E5%BC%82%E8%A1%A8%E8%BE%BE%E5%88%86%E6%9E%90/">

    <span hidden itemprop="author" itemscope itemtype="http://schema.org/Person">
      <meta itemprop="image" content="/codeHub/uploads/themeImg/logo.png">
      <meta itemprop="name" content="">
      <meta itemprop="description" content="">
    </span>

    <span hidden itemprop="publisher" itemscope itemtype="http://schema.org/Organization">
      <meta itemprop="name" content="Code Hub">
    </span>
      <header class="post-header">
        <h1 class="post-title" itemprop="name headline">
          基因差异表达分析
        </h1>

        <div class="post-meta">
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="fa fa-calendar-o"></i>
              </span>
              <span class="post-meta-item-text">Posted on</span>
              

              <time title="Created: 2020-03-22 13:59:13 / Modified: 14:03:41" itemprop="dateCreated datePublished" datetime="2020-03-22T13:59:13+08:00">2020-03-22</time>
            </span>
            <span class="post-meta-item">
              <span class="post-meta-item-icon">
                <i class="fa fa-folder-o"></i>
              </span>
              <span class="post-meta-item-text">In</span>
                <span itemprop="about" itemscope itemtype="http://schema.org/Thing">
                  <a href="/codeHub/categories/RNA-seq/" itemprop="url" rel="index"><span itemprop="name">RNA-seq</span></a>
                </span>
            </span>

          
            <span id="/codeHub/2020/03/22/02%E5%9F%BA%E5%9B%A0%E5%B7%AE%E5%BC%82%E8%A1%A8%E8%BE%BE%E5%88%86%E6%9E%90/" class="post-meta-item leancloud_visitors" data-flag-title="基因差异表达分析" title="Views">
              <span class="post-meta-item-icon">
                <i class="fa fa-eye"></i>
              </span>
              <span class="post-meta-item-text">Views: </span>
              <span class="leancloud-visitors-count"></span>
            </span>
  
  <span class="post-meta-item">
    
      <span class="post-meta-item-icon">
        <i class="fa fa-comment-o"></i>
      </span>
      <span class="post-meta-item-text">Valine: </span>
    
    <a title="valine" href="/codeHub/2020/03/22/02%E5%9F%BA%E5%9B%A0%E5%B7%AE%E5%BC%82%E8%A1%A8%E8%BE%BE%E5%88%86%E6%9E%90/#valine-comments" itemprop="discussionUrl">
      <span class="post-comments-count valine-comment-count" data-xid="/codeHub/2020/03/22/02%E5%9F%BA%E5%9B%A0%E5%B7%AE%E5%BC%82%E8%A1%A8%E8%BE%BE%E5%88%86%E6%9E%90/" itemprop="commentCount"></span>
    </a>
  </span>
  
  

        </div>
      </header>

    
    
    
    <div class="post-body" itemprop="articleBody">

      
        <h1 id="差异表达基因"><a href="#差异表达基因" class="headerlink" title="差异表达基因"></a>差异表达基因</h1><p>​    生物学上不同样本之间的表达差异时服从负二项分布的，RNA-seq中得到的基因表达水平是抽样过程中的一种离散形式。在测得的reads总量一定的情况下，表达水平越高的基因在抽样过程中所占的比例就越高，有些低表达的基因也有可能无法被检测出来。在得到基因的表达量之后，根据实验设计对不同样本之间基因进行差异表达分析</p>
<ol>
<li>同物种、不同组织间的比较</li>
<li>同一物种、同一组织、在不同处理下的比较</li>
<li>同一组织、不同物种间的比较</li>
<li>同一组织在不同时期间的比较</li>
</ol>
<p>​    通过差异表达分析，发现组织特异性、时期特异性、物种特异性的基因表达模式。通过GO功能富集、KEGG分析发现基因在细胞中参与的代谢和具体的功能、基因与基因之间的互作等。</p>
<a id="more"></a>

<h3 id="1-reads计数"><a href="#1-reads计数" class="headerlink" title="1.reads计数"></a>1.reads计数</h3><p>​    使用python包HTseq对统计每个基因比对到的read数</p>
<h4 id="1-1软件安装"><a href="#1-1软件安装" class="headerlink" title="1.1软件安装"></a>1.1软件安装</h4><p>​    非root用户需要使用<code>--user</code>参数</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">pip3 install HTSeq --user</span><br></pre></td></tr></table></figure>

<h4 id="1-2统计基因比对上的read数"><a href="#1-2统计基因比对上的read数" class="headerlink" title="1.2统计基因比对上的read数"></a>1.2统计基因比对上的read数</h4><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">htseq-count  -f bam -r pos -t exon -i gene_id  -m union -q  1_1_5_rmdup.bam genome.gtf &gt;count.txt</span><br></pre></td></tr></table></figure>

<p>​    <strong>命令参数如下</strong>：</p>
<ul>
<li><code>-f | --format</code>设置输入文件格式，默认sam</li>
<li><code>-r | --order</code> 设置输入文件排序方式，默认按照read name排序</li>
<li><code>-s | --stranded</code>是否链特异性建库，默认yes</li>
<li><code>-a | --a</code>设置质量阀值，默认忽略比对质量低于10的read</li>
<li><code>-t | --type</code>对gtf或者gff文件中指定feature计算，默认exon</li>
<li><code>-i | --idattr</code>设置feature id，通常是指第9列中，多个exon共有的gene属性如gene_id</li>
<li><code>-m | --mode default: union</code>设置统计模式</li>
<li><code>-o | --samout</code>输出一个sam文件，比对结果中多一个XF标签比对到的feature id。</li>
<li><code>-q | --quiet</code> 不输出程序运行的状态信息和警告信息</li>
<li><code>-h | --help</code>输出帮助信息。</li>
</ul>
<p><img data-src="https://upload-images.jianshu.io/upload_images/6049898-721e9cbe006abbe3.png" alt="计数模式图"></p>
<h4 id="1-3输出结果"><a href="#1-3输出结果" class="headerlink" title="1.3输出结果"></a>1.3输出结果</h4><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br></pre></td><td class="code"><pre><span class="line">Ghir_A01G000010	11</span><br><span class="line">Ghir_A01G000020	10</span><br><span class="line">Ghir_A01G000030	20</span><br><span class="line">Ghir_A01G000040	139</span><br><span class="line">Ghir_A01G000050	9</span><br><span class="line">Ghir_A01G000060	52</span><br><span class="line">Ghir_A01G000070	68</span><br><span class="line">__no_feature	8716242</span><br><span class="line">__ambiguous	157518</span><br><span class="line">__too_low_aQual	0</span><br><span class="line">__not_aligned	0</span><br><span class="line">__alignment_not_unique	0</span><br></pre></td></tr></table></figure>

<h4 id="1-4批量提交任务"><a href="#1-4批量提交任务" class="headerlink" title="1.4批量提交任务"></a>1.4批量提交任务</h4><figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br></pre></td><td class="code"><pre><span class="line"><span class="keyword">for</span> i <span class="keyword">in</span> `ls `</span><br><span class="line"><span class="keyword">do</span></span><br><span class="line">bsub -J htseqCount -q <span class="string">"smp"</span> -n 1 -R span[hosts=1] -e htseqCount.err -o htseqCount.out <span class="string">"bash htseqcount.sh <span class="variable">$&#123;i&#125;</span>"</span></span><br><span class="line">sleep 1</span><br><span class="line"><span class="keyword">done</span></span><br></pre></td></tr></table></figure>

<h3 id="2-样品无重复"><a href="#2-样品无重复" class="headerlink" title="2.样品无重复"></a>2.样品无重复</h3><p>​    使用DESeq包，对于技术重复作者推荐将两个技术重复的read进行加和后作为样本的read数</p>
<blockquote>
<p> For technical replicates (e. g. when the same library preparation was distributed over multiple lanes of the sequencer), please sum up their counts to get a single column, corresponding to a unique biological replicate.  </p>
</blockquote>
<h4 id="2-1读取原始read数据"><a href="#2-1读取原始read数据" class="headerlink" title="2.1读取原始read数据"></a>2.1读取原始read数据</h4><p>​    其中行名为基因名，列名为样本名</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">	untreated3 untreated4 treated2 treated3</span><br><span class="line">FBgn0000003 0 0 0 1</span><br><span class="line">FBgn0000008 76 70 88 70</span><br><span class="line">FBgn0000014 0 0 0 0</span><br><span class="line">FBgn0000015 1 2 0 0</span><br><span class="line">FBgn0000017 3564 3150 3072 3334</span><br><span class="line">FBgn0000018 245 310 299 308</span><br></pre></td></tr></table></figure>

<h4 id="2-2补充样品分组信息"><a href="#2-2补充样品分组信息" class="headerlink" title="2.2补充样品分组信息"></a>2.2补充样品分组信息</h4><p>​    第一列与第二列属于<code>untreated</code>处理的两个重复</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">condition = factor( c( <span class="string">"untreated"</span>, <span class="string">"untreated"</span>, <span class="string">"treated"</span>, <span class="string">"treated"</span> ) )</span><br></pre></td></tr></table></figure>

<h4 id="2-3将分组信息与read表进行合并"><a href="#2-3将分组信息与read表进行合并" class="headerlink" title="2.3将分组信息与read表进行合并"></a>2.3将分组信息与read表进行合并</h4><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">&gt; <span class="keyword">library</span>( <span class="string">"DESeq"</span> )</span><br><span class="line">&gt; cds = newCountDataSet( countTable, condition )</span><br></pre></td></tr></table></figure>

<h4 id="2-4对不同处理进行标准化"><a href="#2-4对不同处理进行标准化" class="headerlink" title="2.4对不同处理进行标准化"></a>2.4对不同处理进行标准化</h4><p>​    通过<code>estimateSizeFactors( cds )</code>函数来计算不同处理间测序深度是否存在较大的差异</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">cds=estimateSizeFactors( cds )</span><br><span class="line">counts(cds, normalized = <span class="literal">TRUE</span>)</span><br></pre></td></tr></table></figure>

<h4 id="2-5估计离散度"><a href="#2-5估计离散度" class="headerlink" title="2.5估计离散度"></a>2.5估计离散度</h4><ul>
<li>这里由于没有重复需要使用<code>method= &quot;blind&quot;, sharingMode = &quot;fit-only&quot;</code>参数</li>
</ul>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br></pre></td><td class="code"><pre><span class="line">cds &lt;- estimateDispersions(cds, method = <span class="string">"blind"</span>, sharingMode = <span class="string">"fit-only"</span>)</span><br></pre></td></tr></table></figure>

<h4 id="2-6差异分析"><a href="#2-6差异分析" class="headerlink" title="2.6差异分析"></a>2.6差异分析</h4><p>​    后两个参数是指定需要比较的样品</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">res = nbinomTest( cds, <span class="string">"untreated"</span>, <span class="string">"treated"</span> )</span><br><span class="line"><span class="comment">## 绘制数据的分布情况</span></span><br><span class="line">plotMA(res)</span><br></pre></td></tr></table></figure>

<p><img data-src="https://s1.ax1x.com/2020/03/17/8t6jXV.png" alt="数据分布"></p>
<h4 id="2-7输出结果"><a href="#2-7输出结果" class="headerlink" title="2.7输出结果"></a>2.7输出结果</h4><ul>
<li><code>id feature</code> identifier baseMean mean normalised counts, averaged over all samples from both conditions +</li>
<li><code>baseMeanA</code> mean normalised counts from condition A </li>
<li><code>baseMeanB</code>mean normalised counts from condition B foldChange </li>
<li><code>fold change</code> from condition A to B </li>
<li><code>log2FoldChange</code> the logarithm (to basis 2) of the fold change </li>
<li><code>pval</code>p value for the statistical significance of this change </li>
<li><code>padj</code>p value adjusted for multiple testing with the Benjamini-Hochberg procedure (see the R function p.adjust), which controls false discovery rate (FDR)  </li>
</ul>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">id baseMean baseMeanA baseMeanB foldChange log2FoldChange pval padj</span><br><span class="line">1 FBgn0000003 0.224 0.00 0.449 Inf Inf 1.000 1.000</span><br><span class="line">2 FBgn0000008 76.296 78.16 74.436 0.952 -0.0704 0.835 1.000</span><br><span class="line">3 FBgn0000014 0.000 0.00 0.000 NaN NaN NA NA</span><br><span class="line">4 FBgn0000015 0.781 1.56 0.000 0.000 -Inf 0.416 1.000</span><br><span class="line">5 FBgn0000017 3298.682 3599.47 2997.890 0.833 -0.2638 0.241 0.881</span><br><span class="line">6 FBgn0000018 289.031 293.68 284.385 0.968 -0.0464 0.757 1.000</span><br></pre></td></tr></table></figure>

<h4 id="2-8筛选差异表达基因"><a href="#2-8筛选差异表达基因" class="headerlink" title="2.8筛选差异表达基因"></a>2.8筛选差异表达基因</h4><p>​    没有重复的样根据p-value来筛选差异表达的基因意义不大，所以直接对输出的结果用awk进行筛选。筛选的时候有三种情况</p>
<ul>
<li>两个样都有read比对上</li>
<li>两个样中有一个样是没有read比对上，这种情况会使的log2foldcahnge为inf</li>
<li>两个样中比对到的read都为0</li>
</ul>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br></pre></td><td class="code"><pre><span class="line">awk -F <span class="string">"\t"</span> <span class="string">'NR&gt;=2&amp;&amp;$6!="Inf"&amp;&amp;$6!="NA"&amp;&amp;$6&gt;=1&#123;print $1"\tup"&#125;NR&gt;=2&amp;&amp;$6!="Inf"&amp;&amp;$6!="NA"&amp;&amp;$6&lt;=-1&#123;print $1"\tdown"&#125;'</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## 第三种情况相当于没有差异表达，不用考虑</span></span><br></pre></td></tr></table></figure>



<h3 id="3-样品有重复"><a href="#3-样品有重复" class="headerlink" title="3.样品有重复"></a>3.样品有重复</h3><p>​    推荐使用DESeq2包</p>
<h3 id="参考"><a href="#参考" class="headerlink" title="参考"></a>参考</h3><ol>
<li>HTSeq仓库  <a href="https://github.com/simon-anders/htseq" target="_blank" rel="noopener">https://github.com/simon-anders/htseq</a> </li>
<li>HTSeq使用文档  <a href="https://htseq.readthedocs.io/en/release_0.11.1/" target="_blank" rel="noopener">https://htseq.readthedocs.io/en/release_0.11.1/</a> </li>
<li>HTSeq使用文档  <a href="https://www.cnblogs.com/triple-y/p/9338890.html" target="_blank" rel="noopener">https://www.cnblogs.com/triple-y/p/9338890.html</a> </li>
<li>DESeq使用文档  <a href="https://bioconductor.org/packages/release/bioc/vignettes/DESeq/inst/doc/DESeq.pdf" target="_blank" rel="noopener">https://bioconductor.org/packages/release/bioc/vignettes/DESeq/inst/doc/DESeq.pdf</a> </li>
</ol>

    </div>

    
    
    
        

<div>
<ul class="post-copyright">
  <li class="post-copyright-author">
    <strong>Post author:  </strong>zpliu
  </li>
  <li class="post-copyright-link">
    <strong>Post link: </strong>
    <a href="https://zpliu1126.github.io/codeHub/2020/03/22/02%E5%9F%BA%E5%9B%A0%E5%B7%AE%E5%BC%82%E8%A1%A8%E8%BE%BE%E5%88%86%E6%9E%90/" title="基因差异表达分析">https://zpliu1126.github.io/codeHub/2020/03/22/02基因差异表达分析/</a>
  </li>
  <li class="post-copyright-license">
    <strong>Copyright Notice:  </strong>All articles in this blog are licensed under <a href="https://creativecommons.org/licenses/by-nc-sa/4.0/zh-cn" rel="noopener" target="_blank"><i class="fa fa-fw fa-creative-commons"></i>BY-NC-SA</a> unless stating additionally.
  </li>
</ul>
</div>


      <footer class="post-footer">

        


        
    <div class="post-nav">
      <div class="post-nav-item">
    <a href="/codeHub/2020/03/21/01hexo%E6%9C%8D%E5%8A%A1%E6%90%AD%E5%BB%BA/" rel="prev" title="hexo服务搭建">
      <i class="fa fa-chevron-left"></i> hexo服务搭建
    </a></div>
      <div class="post-nav-item">
    <a href="/codeHub/2020/03/22/03circos%E4%BD%9C%E5%9B%BE%E5%9F%BA%E7%A1%80/" rel="next" title="circos作图基础">
      circos作图基础 <i class="fa fa-chevron-right"></i>
    </a></div>
    </div>
      </footer>
    
  </article>
  
  
  

  </div>


          </div>
          
    <div class="comments" id="valine-comments"></div>

<script>
  window.addEventListener('tabs:register', () => {
    let { activeClass } = CONFIG.comments;
    if (CONFIG.comments.storage) {
      activeClass = localStorage.getItem('comments_active') || activeClass;
    }
    if (activeClass) {
      let activeTab = document.querySelector(`a[href="#comment-${activeClass}"]`);
      if (activeTab) {
        activeTab.click();
      }
    }
  });
  if (CONFIG.comments.storage) {
    window.addEventListener('tabs:click', event => {
      if (!event.target.matches('.tabs-comment .tab-content .tab-pane')) return;
      let commentClass = event.target.classList[1];
      localStorage.setItem('comments_active', commentClass);
    });
  }
</script>

        </div>
          
  
  <div class="toggle sidebar-toggle">
    <span class="toggle-line toggle-line-first"></span>
    <span class="toggle-line toggle-line-middle"></span>
    <span class="toggle-line toggle-line-last"></span>
  </div>

  <aside class="sidebar">
    <div class="sidebar-inner">

      <ul class="sidebar-nav motion-element">
        <li class="sidebar-nav-toc">
          文章目录
        </li>
        <li class="sidebar-nav-overview">
          站点概括
        </li>
      </ul>

      <!--noindex-->
      <div class="post-toc-wrap sidebar-panel">
          <div class="post-toc motion-element"><ol class="nav"><li class="nav-item nav-level-1"><a class="nav-link" href="#差异表达基因"><span class="nav-number">1.</span> <span class="nav-text">差异表达基因</span></a><ol class="nav-child"><li class="nav-item nav-level-3"><a class="nav-link" href="#1-reads计数"><span class="nav-number">1.0.1.</span> <span class="nav-text">1.reads计数</span></a><ol class="nav-child"><li class="nav-item nav-level-4"><a class="nav-link" href="#1-1软件安装"><span class="nav-number">1.0.1.1.</span> <span class="nav-text">1.1软件安装</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#1-2统计基因比对上的read数"><span class="nav-number">1.0.1.2.</span> <span class="nav-text">1.2统计基因比对上的read数</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#1-3输出结果"><span class="nav-number">1.0.1.3.</span> <span class="nav-text">1.3输出结果</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#1-4批量提交任务"><span class="nav-number">1.0.1.4.</span> <span class="nav-text">1.4批量提交任务</span></a></li></ol></li><li class="nav-item nav-level-3"><a class="nav-link" href="#2-样品无重复"><span class="nav-number">1.0.2.</span> <span class="nav-text">2.样品无重复</span></a><ol class="nav-child"><li class="nav-item nav-level-4"><a class="nav-link" href="#2-1读取原始read数据"><span class="nav-number">1.0.2.1.</span> <span class="nav-text">2.1读取原始read数据</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#2-2补充样品分组信息"><span class="nav-number">1.0.2.2.</span> <span class="nav-text">2.2补充样品分组信息</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#2-3将分组信息与read表进行合并"><span class="nav-number">1.0.2.3.</span> <span class="nav-text">2.3将分组信息与read表进行合并</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#2-4对不同处理进行标准化"><span class="nav-number">1.0.2.4.</span> <span class="nav-text">2.4对不同处理进行标准化</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#2-5估计离散度"><span class="nav-number">1.0.2.5.</span> <span class="nav-text">2.5估计离散度</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#2-6差异分析"><span class="nav-number">1.0.2.6.</span> <span class="nav-text">2.6差异分析</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#2-7输出结果"><span class="nav-number">1.0.2.7.</span> <span class="nav-text">2.7输出结果</span></a></li><li class="nav-item nav-level-4"><a class="nav-link" href="#2-8筛选差异表达基因"><span class="nav-number">1.0.2.8.</span> <span class="nav-text">2.8筛选差异表达基因</span></a></li></ol></li><li class="nav-item nav-level-3"><a class="nav-link" href="#3-样品有重复"><span class="nav-number">1.0.3.</span> <span class="nav-text">3.样品有重复</span></a></li><li class="nav-item nav-level-3"><a class="nav-link" href="#参考"><span class="nav-number">1.0.4.</span> <span class="nav-text">参考</span></a></li></ol></li></ol></li></ol></div>
      </div>
      <!--/noindex-->

      <div class="site-overview-wrap sidebar-panel">
        <div class="site-author motion-element" itemprop="author" itemscope itemtype="http://schema.org/Person">
    <img class="site-author-image" itemprop="image" alt=""
      src="/codeHub/uploads/themeImg/logo.png">
  <p class="site-author-name" itemprop="name"></p>
  <div class="site-description" itemprop="description"></div>
</div>
<div class="site-state-wrap motion-element">
  <nav class="site-state">
      <div class="site-state-item site-state-posts">
          <a href="/codeHub/archives">
          <span class="site-state-item-count">3</span>
          <span class="site-state-item-name">文章</span>
        </a>
      </div>
      <div class="site-state-item site-state-categories">
            <a href="/codeHub/categories/">
        <span class="site-state-item-count">2</span>
        <span class="site-state-item-name">分类</span></a>
      </div>
      <div class="site-state-item site-state-tags">
            <a href="/codeHub/tags/">
        <span class="site-state-item-count">2</span>
        <span class="site-state-item-name">标签</span></a>
      </div>
  </nav>
</div>
  <div class="links-of-author motion-element">
      <span class="links-of-author-item">
        <a href="https://github.com/zpliu1126" title="GitHub → https:&#x2F;&#x2F;github.com&#x2F;zpliu1126" rel="noopener" target="_blank"><i class="fa fa-fw fa-github"></i>GitHub</a>
      </span>
      <span class="links-of-author-item">
        <a href="/codeHub/1944532210@qq.com" title="E-Mail → 1944532210@qq.com"><i class="fa fa-fw fa-envelope"></i>E-Mail</a>
      </span>
  </div>
  <div class="cc-license motion-element" itemprop="license">
    <a href="https://creativecommons.org/licenses/by-nc-sa/4.0/zh-cn" class="cc-opacity" rel="noopener" target="_blank"><img src="/codeHub/images/cc-by-nc-sa.svg" alt="Creative Commons"></a>
  </div>


  <div class="links-of-blogroll motion-element">
    <div class="links-of-blogroll-title">
      <i class="fa fa-fw fa-link"></i>
      友情链接
    </div>
    <ul class="links-of-blogroll-list">
        <li class="links-of-blogroll-item">
          <a href="http://cotton.hzau.edu.cn/" title="http:&#x2F;&#x2F;cotton.hzau.edu.cn&#x2F;" rel="noopener" target="_blank">GCGI</a>
        </li>
    </ul>
  </div>

      </div>
        <div class="back-to-top motion-element">
          <i class="fa fa-arrow-up"></i>
          <span>0%</span>
        </div>

    </div>
  </aside>
  <div id="sidebar-dimmer"></div>


      </div>
    </main>

    <footer class="footer">
      <div class="footer-inner">
        

        

<div class="copyright">
  
  &copy; 2019 – 
  <span itemprop="copyrightYear">2020</span>
  <span class="with-love">
    <i class="fa fa-user"></i>
  </span>
  <span class="author" itemprop="copyrightHolder">codeHub</span>
</div>

        






<script pjax>
  (function() {
    function leancloudSelector(url) {
      url = encodeURI(url);
      return document.getElementById(url).querySelector('.leancloud-visitors-count');
    }

    function addCount(Counter) {
      var visitors = document.querySelector('.leancloud_visitors');
      var url = decodeURI(visitors.id);
      var title = visitors.dataset.flagTitle;

      Counter('get', '/classes/Counter?where=' + encodeURIComponent(JSON.stringify({ url })))
        .then(response => response.json())
        .then(({ results }) => {
          if (results.length > 0) {
            var counter = results[0];
            leancloudSelector(url).innerText = counter.time + 1;
            Counter('put', '/classes/Counter/' + counter.objectId, { time: { '__op': 'Increment', 'amount': 1 } })
              .catch(error => {
                console.error('Failed to save visitor count', error);
              });
          } else {
              leancloudSelector(url).innerText = 'Counter not initialized! More info at console err msg.';
              console.error('ATTENTION! LeanCloud counter has security bug, see how to solve it here: https://github.com/theme-next/hexo-leancloud-counter-security. \n However, you can still use LeanCloud without security, by setting `security` option to `false`.');
            
          }
        })
        .catch(error => {
          console.error('LeanCloud Counter Error', error);
        });
    }

    function showTime(Counter) {
      var visitors = document.querySelectorAll('.leancloud_visitors');
      var entries = [...visitors].map(element => {
        return decodeURI(element.id);
      });

      Counter('get', '/classes/Counter?where=' + encodeURIComponent(JSON.stringify({ url: { '$in': entries } })))
        .then(response => response.json())
        .then(({ results }) => {
          for (let url of entries) {
            let target = results.find(item => item.url === url);
            leancloudSelector(url).innerText = target ? target.time : 0;
          }
        })
        .catch(error => {
          console.error('LeanCloud Counter Error', error);
        });
    }

    let { app_id, app_key, server_url } = {"enable":true,"app_id":"d4VOUvNSU5WGIC3AjjM8UArx-gzGzoHsz","app_key":"5n9Xu8XStUzmAKyncMgSp0px","server_url":"https://d4vouvns.lc-cn-n1-shared.com","security":true};
    function fetchData(api_server) {
      var Counter = (method, url, data) => {
        return fetch(`${api_server}/1.1${url}`, {
          method,
          headers: {
            'X-LC-Id'     : app_id,
            'X-LC-Key'    : app_key,
            'Content-Type': 'application/json',
          },
          body: JSON.stringify(data)
        });
      };
      if (CONFIG.page.isPost) {
        if (CONFIG.hostname !== location.hostname) return;
        addCount(Counter);
      } else if (document.querySelectorAll('.post-title-link').length >= 1) {
        showTime(Counter);
      }
    }

    let api_server = app_id.slice(-9) !== '-MdYXbMMI' ? server_url : `https://${app_id.slice(0, 8).toLowerCase()}.api.lncldglobal.com`;

    if (api_server) {
      fetchData(api_server);
    } else {
      fetch('https://app-router.leancloud.cn/2/route?appId=' + app_id)
        .then(response => response.json())
        .then(({ api_server }) => {
          fetchData('https://' + api_server);
        });
    }
  })();
</script>


      </div>
    </footer>
  </div>

  
  <script src="/codeHub/lib/anime.min.js"></script>
  <script src="https://cdn.jsdelivr.net/npm/pjax@0.2.8/pjax.min.js"></script>
  <script src="//cdn.jsdelivr.net/npm/jquery@3/dist/jquery.min.js"></script>
  <script src="//cdn.jsdelivr.net/gh/fancyapps/fancybox@3/dist/jquery.fancybox.min.js"></script>
  <script src="//cdn.jsdelivr.net/npm/lozad@1/dist/lozad.min.js"></script>
  <script src="/codeHub/lib/velocity/velocity.min.js"></script>
  <script src="/codeHub/lib/velocity/velocity.ui.min.js"></script>

<script src="/codeHub/js/utils.js"></script>

<script src="/codeHub/js/motion.js"></script>


<script src="/codeHub/js/schemes/pisces.js"></script>


<script src="/codeHub/js/next-boot.js"></script>

<script src="/codeHub/js/bookmark.js"></script>

  <script>
var pjax = new Pjax({
  selectors: [
    'head title',
    '#page-configurations',
    '.content-wrap',
    '.post-toc-wrap',
    '.languages',
    '#pjax'
  ],
  switches: {
    '.post-toc-wrap': Pjax.switches.innerHTML
  },
  analytics: false,
  cacheBust: false,
  scrollTo : !CONFIG.bookmark.enable
});

window.addEventListener('pjax:success', () => {
  document.querySelectorAll('script[pjax], script#page-configurations, #pjax script').forEach(element => {
    var code = element.text || element.textContent || element.innerHTML || '';
    var parent = element.parentNode;
    parent.removeChild(element);
    var script = document.createElement('script');
    if (element.id) {
      script.id = element.id;
    }
    if (element.className) {
      script.className = element.className;
    }
    if (element.type) {
      script.type = element.type;
    }
    if (element.src) {
      script.src = element.src;
      // Force synchronous loading of peripheral JS.
      script.async = false;
    }
    if (element.getAttribute('pjax') !== null) {
      script.setAttribute('pjax', '');
    }
    if (code !== '') {
      script.appendChild(document.createTextNode(code));
    }
    parent.appendChild(script);
  });
  NexT.boot.refresh();
  // Define Motion Sequence & Bootstrap Motion.
  if (CONFIG.motion.enable) {
    NexT.motion.integrator
      .init()
      .add(NexT.motion.middleWares.subMenu)
      .add(NexT.motion.middleWares.postList)
      .bootstrap();
  }
  NexT.utils.updateSidebarPosition();
});
</script>




  




  
<script src="/codeHub/js/local-search.js"></script>













    <div id="pjax">
  

  


<script>
NexT.utils.loadComments(document.querySelector('#valine-comments'), () => {
  NexT.utils.getScript('//unpkg.com/valine/dist/Valine.min.js', () => {
    var GUEST = ['nick', 'mail', 'link'];
    var guest = 'nick,mail,link';
    guest = guest.split(',').filter(item => {
      return GUEST.includes(item);
    });
    new Valine({
      el         : '#valine-comments',
      verify     : false,
      notify     : false,
      appId      : 'd4VOUvNSU5WGIC3AjjM8UArx-gzGzoHsz',
      appKey     : '5n9Xu8XStUzmAKyncMgSp0px',
      placeholder: "有任何问题欢迎留言或者邮件告知",
      avatar     : 'mm',
      meta       : guest,
      pageSize   : '10' || 10,
      visitor    : false,
      lang       : 'zh-cn' || 'zh-cn',
      path       : location.pathname,
      recordIP   : false,
      serverURLs : ''
    });
  }, window.Valine);
});
</script>

    </div>
</body>
</html>
